File: server/modules/git/worktree-list.ts
import { existsSync } from "node:fs";
import { runGit } from "./exec.js";

export interface GitWorktreeEntry {
  path: string;
  head?: string;
  branch?: string;
  gitDir?: string;
}

export function parseWorktreePorcelain(output: string): GitWorktreeEntry[] {
  const entries: GitWorktreeEntry[] = [];
  let current: Partial<GitWorktreeEntry> = {};

  const flush = () => {
    if (current.path) {
      entries.push({
        path: current.path,
        head: current.head,
        branch: current.branch?.replace(/^refs\/heads\//, ""),
        gitDir: current.gitDir,
      });
    }
    current = {};
  };

  for (const line of output.split("\n")) {
    if (!line.trim()) {
      flush();
      continue;
    }
    if (line.startsWith("worktree ")) {
      current.path = line.slice("worktree ".length).trim();
    } else if (line.startsWith("HEAD ")) {
      current.head = line.slice("HEAD ".length).trim();
    } else if (line.startsWith("branch ")) {
      current.branch = line.slice("branch ".length).trim();
    } else if (line.startsWith("gitdir ")) {
      current.gitDir = line.slice("gitdir ".length).trim();
    }
  }
  flush();
  return entries;
}

export function listWorktrees(repoPath: string): GitWorktreeEntry[] {
  const out = runGit(repoPath, ["worktree", "list", "--porcelain"]);
  return parseWorktreePorcelain(out);
}

export function listBranches(repoPath: string): string[] {
  const out = runGit(repoPath, [
    "branch",
    "--format=%(refname:short)",
  ]);
  return out
    .split("\n")
    .map((b) => b.trim())
    .filter(Boolean);
}

export function getDefaultBranch(repoPath: string): string {
  try {
    const sym = runGit(repoPath, ["symbolic-ref", "refs/remotes/origin/HEAD"]);
    const match = sym.match(/refs\/remotes\/origin\/(.+)/);
    if (match?.[1]) return match[1];
  } catch {
    // fall through
  }
  try {
    return runGit(repoPath, ["branch", "--show-current"]) || "main";
  } catch {
    const branches = listBranches(repoPath);
    return branches[0] ?? "main";
  }
}

export function branchExists(repoPath: string, branch: string): boolean {
  return listBranches(repoPath).includes(branch);
}

export function worktreePathExists(path: string): boolean {
  return existsSync(path);
}

File: server-rs/src/git/worktree_list.rs
use super::exec::{run_git, GitError};
use std::path::Path;

#[derive(Debug, Clone)]
pub struct GitWorktreeEntry {
    pub path: String,
    pub head: Option<String>,
    pub branch: Option<String>,
    pub git_dir: Option<String>,
}

pub fn parse_worktree_porcelain(output: &str) -> Vec<GitWorktreeEntry> {
    let mut entries = Vec::new();
    let mut current: Option<GitWorktreeEntry> = None;

    let flush = |current: &mut Option<GitWorktreeEntry>, entries: &mut Vec<GitWorktreeEntry>| {
        if let Some(entry) = current.take() {
            if !entry.path.is_empty() {
                entries.push(entry);
            }
        }
    };

    for line in output.lines() {
        if line.trim().is_empty() {
            flush(&mut current, &mut entries);
            continue;
        }
        if let Some(rest) = line.strip_prefix("worktree ") {
            flush(&mut current, &mut entries);
            current = Some(GitWorktreeEntry {
                path: rest.trim().to_string(),
                head: None,
                branch: None,
                git_dir: None,
            });
        } else if let Some(entry) = current.as_mut() {
            if let Some(rest) = line.strip_prefix("HEAD ") {
                entry.head = Some(rest.trim().to_string());
            } else if let Some(rest) = line.strip_prefix("branch ") {
                let branch = rest.trim().trim_start_matches("refs/heads/").to_string();
                entry.branch = Some(branch);
            } else if let Some(rest) = line.strip_prefix("gitdir ") {
                entry.git_dir = Some(rest.trim().to_string());
            }
        }
    }
    flush(&mut current, &mut entries);
    entries
}

pub fn list_worktrees(repo_path: &str) -> Result<Vec<GitWorktreeEntry>, GitError> {
    let out = run_git(repo_path, &["worktree", "list", "--porcelain"], None)?;
    Ok(parse_worktree_porcelain(&out))
}

pub fn list_branches(repo_path: &str) -> Result<Vec<String>, GitError> {
    let out = run_git(
        repo_path,
        &["branch", "--format=%(refname:short)"],
        None,
    )?;
    Ok(out
        .lines()
        .map(str::trim)
        .filter(|b| !b.is_empty())
        .map(str::to_string)
        .collect())
}

pub fn get_default_branch(repo_path: &str) -> String {
    if let Ok(sym) = run_git(
        repo_path,
        &["symbolic-ref", "refs/remotes/origin/HEAD"],
        None,
    ) {
        if let Some(branch) = sym.strip_prefix("refs/remotes/origin/") {
            if !branch.is_empty() {
                return branch.to_string();
            }
        }
    }

    if let Ok(current) = run_git(repo_path, &["branch", "--show-current"], None) {
        if !current.is_empty() {
            return current;
        }
    }

    list_branches(repo_path)
        .ok()
        .and_then(|branches| branches.into_iter().next())
        .unwrap_or_else(|| "main".to_string())
}

pub fn branch_exists(repo_path: &str, branch: &str) -> Result<bool, GitError> {
    Ok(list_branches(repo_path)?.iter().any(|b| b == branch))
}

pub fn worktree_path_exists(path: &str) -> bool {
    Path::new(path).exists()
}
